### PR TITLE
Reduce descriptor size if remaining bytes is less than request size

### LINKED ISSUE #595: Files over 128k not of size (N * 128k) create a "file too large for filesystem" error when using rdma
It has been a while since I used rdma since I switched to windows 11, trying to get it working again.

### Issue
See title, files with offsets other than blocks of 128k will not transfer properly.

### Setup
Version of kmod: git master
Version of tools: git master
kernel: 6.4.7-xanmod with patch for mellanox adapter to be recognised as rdma device
Changes from default: 8 rdma connection instead of 2
Adapter: mellanox ConnectX-4 EN 100gbit, winof2 on windows, inbox on linux
Client OS: Windows 11 for workstations
Server OS: Debian bookworm

### Expected behaviour
Files copy normally

### To reproduce
```
dd if=/dev/urandom of=127k.bin bs=1024 count=127
dd if=/dev/urandom of=128k.bin bs=1024 count=128
dd if=/dev/urandom of=129k.bin bs=1024 count=129
```
copy files from ksmbd location (linux, zfs) to local storage (windows, ntfs)

### Result
image

Also tried from ext4 partition: same problem to avoid zfs being the issue.

Afterwards I created a 256k file since my testfile of 8gb copied fine:
```
dd if=/dev/urandom of=256k.bin bs=1024 count=256
```
Which worked. (not included in logs)

It look like there is some issue with non 128k sized files (offset issue?). The part which stands out:
```
[48394.612054] ksmbd: smb_direct: RDMA write, len 0x20000, needed credits 0x1
[48394.612572] ksmbd: filename 129k.bin, offset 131072, len 131072
[48394.614189] ksmbd: nbytes 1024, offset 132096 mincount 0
[48394.614585] ksmbd: Failed to process 8 [-22]
```

Attached are the logs (`ksmbd.control -d all`), somewhat sanitized

Additional notes:
Sometimes windows acts like it is transfering the file twice.
[result.log](https://github.com/cifsd-team/ksmbd/files/12208943/result.log)

### PR DESCRIPTION
**Pull request by consp**
Fixes #595

### CODE REVIEW

File: smb2pdu.c [hunk 1]
```diff
@@ -6792,7 +6792,13 @@ static ssize_t smb2_read_rdma_channel(struct ksmbd_work *work,
...
+    struct smb2_buffer_desc_v1 *desc;
 
+    /* set descriptor lenght to nbytes if less than request size */
+    if (length < req->Length) {
+        desc = (struct smb2_buffer_desc_v1 *)((char *)req + le16_to_cpu(req->ReadChannelInfoOffset));
+        desc->length = cpu_to_le32(length);
+    }
 	err = ksmbd_conn_rdma_write(work->conn, data_buf, length,
 				    (struct smb2_buffer_desc_v1 *)
 				    ((char *)req + le16_to_cpu(req->ReadChannelInfoOffset)),
```
Comment (namjaejeon) [inline smb2pdu.c:6804]: Please change it to desc.

File: smb2pdu.c [hunk 2]
```diff
@@ -6792,7 +6792,13 @@ static ssize_t smb2_read_rdma_channel(struct ksmbd_work *work,
 				      size_t length)
 {
 	int err;
+    struct smb2_buffer_desc_v1 *desc;
 
+    /* set descriptor lenght to nbytes if less than request size */
+    if (length < req->Length) {
+        desc = (struct smb2_buffer_desc_v1 *)((char *)req + le16_to_cpu(req->ReadChannelInfoOffset));
```
Comment (namjaejeon) [inline smb2pdu.c:6799]: move it to out of if condition.

```
desc = (struct smb2_buffer_desc_v1 *)((char *)req + le16_to_cpu(req->ReadChannelInfoOffset));
/* set descriptor lenght to nbytes if less than request size */
if (length < req->Length)

**Comment by namjaejeon** [comment]
Please update the subject to "ksmbd: reduce descriptor size if remaining bytes is less than request size".
And need to add signed-off-by tag. you can create the patch using git.

**Comment by namjaejeon** [comment]
Have you ever sent the patch to the mailing list ? ksmbd patch should be sent to the mailing list(linux-cifs@vger.kernel.org)

**Comment by namjaejeon** [comment]
Can you update the patch description ? You can add the description of ISSUE and how this patch fix it.

**Comment by consp** [comment]
Yes, I'll add the changes this weekend (This is on my private time). No I have never send a patch to the kernel mailing list since it look like too much effort to do on your own time but I'll have a look.

**Comment by namjaejeon** [comment]
Okay, I will directly update your patch and send the patch to mailing list. Let me know your full name and mail address to add signed-off-by tag.

**Comment by namjaejeon** [comment]
>  Let me know your full name and mail address to add signed-off-by tag.

@consp Ping?

**Comment by namjaejeon** [comment]
@consp I have reproduced it with cifs client, I will check it more. thanks.

**Comment by namjaejeon** [comment]
Applied updated patch. Let me know if you have the issue.

**Pull request closed without merging**